File: GridView/EFCodeFirstInHierarchy/RadGridViewEFCodeFirstInHierarchyVB/RadGridViewEFCodeFirst.Data/RadGridViewEFCodeFirstData.vb
Imports RadGridViewEFCodeFirst.Models

Public Class RadGridViewEFCodeFirstData
    Implements IRadGridViewEFCodeFirstData

    Private context As IRadGridViewEFCodeFirstContext
    Private repositories As IDictionary(Of Type, Object)

    Public Sub New()
        Me.New(New RadGridViewEFCodeFirstContext())
    End Sub

    Public Sub New(context As IRadGridViewEFCodeFirstContext)
        Me.context = context
        Me.repositories = New Dictionary(Of Type, Object)()
    End Sub

    Public ReadOnly Property Orders() As IGenericRepository(Of Order) Implements IRadGridViewEFCodeFirstData.Orders
        Get
            Return Me.GetRepository(Of Order)()
        End Get
    End Property

    Public ReadOnly Property OrderTypes() As IGenericRepository(Of OrderType) Implements IRadGridViewEFCodeFirstData.OrderTypes
        Get
            Return Me.GetRepository(Of OrderType)()
        End Get
    End Property

    Public ReadOnly Property Shippers() As IGenericRepository(Of Shipper) Implements IRadGridViewEFCodeFirstData.Shippers
        Get
            Return Me.GetRepository(Of Shipper)()
        End Get
    End Property

    Public Sub SaveChanges() Implements IRadGridViewEFCodeFirstData.SaveChanges
        Me.context.SaveChanges()
    End Sub

    Private Function GetRepository(Of T As Class)() As IGenericRepository(Of T)
        Dim typeOfModel = GetType(T)
        If Not Me.repositories.ContainsKey(typeOfModel) Then
            Dim type = GetType(RadGridViewEFCodeFirstRepository(Of T))

            Me.repositories.Add(typeOfModel, Activator.CreateInstance(type, Me.context))
        End If

        Return DirectCast(Me.repositories(typeOfModel), IGenericRepository(Of T))
    End Function
End Class

File: GridView/EFCodeFirstInHierarchy/RadGridViewEFCodeFirstInHierarchy/RadGridViewEFCodeFirst.Data/RadGridViewEFCodeFirstData.cs
using System;
using System.Collections.Generic;
using System.Linq;

using RadGridViewEFCodeFirst.Data.Contracts;
using RadGridViewEFCodeFirst.Models;

namespace RadGridViewEFCodeFirst.Data
{
    public class RadGridViewEFCodeFirstData : IRadGridViewEFCodeFirstData
    {
        private IRadGridViewEFCodeFirstContext context;
        private IDictionary<Type, object> repositories;

        public RadGridViewEFCodeFirstData()
            : this(new RadGridViewEFCodeFirstContext())
        {
        }

        public RadGridViewEFCodeFirstData(IRadGridViewEFCodeFirstContext context)
        {
            this.context = context;
            this.repositories = new Dictionary<Type, object>();
        }

        public IGenericRepository<Order> Orders
        {
            get
            {
                return this.GetRepository<Order>();
            }
        }

        public IGenericRepository<OrderType> OrderTypes
        {
            get
            {
                return this.GetRepository<OrderType>();
            }
        }

        public IGenericRepository<Shipper> Shippers
        {
            get
            {
                return this.GetRepository<Shipper>();
            }
        }

        public void SaveChanges()
        {
            this.context.SaveChanges();
        }

        private IGenericRepository<T> GetRepository<T>() where T : class
        {
            var typeOfModel = typeof(T);
            if (!this.repositories.ContainsKey(typeOfModel))
            {
                Type type = typeof(RadGridViewEFCodeFirstRepository<T>);
                this.repositories.Add(typeOfModel, Activator.CreateInstance(type, this.context));
            }

            return (IGenericRepository<T>)this.repositories[typeOfModel];
        }
    }
}
